### ISSUE TITLE
[Risk]: riskAgentMemoryPoisoning

### ISSUE THREAD

**Issue by jo5iah** [cosai-oasis/secure-ai-tooling#524]
### Risk Title*

Memory Poisoning

### Short Description*

Poisoning of an agent's *persistent long-term memory* so that future sessions inherit malicious state. Memory here means any store the agent writes to in one turn and reads back in a later one, whatever the substrate: a vector store, a scratchpad file, a database row, a project instruction file, or a file the agent edits in a repository it also reads from. The read-after-write-across-turns property is what creates the surface, because it is what lets an implant outlive the session that planted it.

### Long Description*

The Telemetry SIG uncovered this gap in producing https://github.com/cosai-oasis/ws2-defenders/blob/main/telemetry/CoSAI-AI-Telemetry-RFC.md

No entry in the post-#507 corpus names `componentMemory` as the poisoned asset.

### Risk Category*

risksSupplyChainAndDevelopment

### Applicable Personas*

- [ ] personaModelProvider
- [ ] personaDataProvider
- [ ] personaPlatformProvider
- [ ] personaModelServing
- [x] personaAgenticProvider
- [ ] personaApplicationDeveloper
- [x] personaEndUser

### Applicable Controls*

This will require a new controlAgentMemoryIntegrity, for which a related issue will be created.

### Examples (Optional)

Three catalogued incidents, all with the agent's own memory as the persistence mechanism:

- **MINJA (Memory Injection Attack)**, from the CoSAI Workstream 2 *AI Incident Response Framework* case studies (<https://github.com/cosai-oasis/ws2-defenders/blob/main/incident-response/AI-Incident-Response.md>). Benign-looking queries induce the agent to autonomously generate and persist malicious reasoning in memory. MITRE ATLAS (<https://atlas.mitre.org/>): `AML.T0051` LLM Prompt Injection, `AML.T0070` RAG Poisoning, `AML.T0059` Erode Dataset Integrity, `AML.T0061` LLM Prompt Self-Replication, `AML.T0067` LLM Trusted Output Components Manipulation.
- **AGENTPOISON**, same CoSAI WS2 source. Optimized trigger-based adversarial queries poison agent memory and RAG. ATLAS: `AML.T0070`, `AML.T0020` Poison Training Data, `AML.T0043` Craft Adversarial Data.
- **Agent Corruption**, case AOC-10 from Shapira, N., Wendler, C., Yen, A., et al., *Agents of Chaos*, arXiv:2602.20021 (2026), <https://arxiv.org/abs/2602.20021> (interactive log: <https://agentsofchaos.baulab.info/>). Indirect injection via an externally editable Gist "constitution" linked from memory; the injected content caused the agent to shut down peers, ban users, and send unauthorized email. ATLAS: `AML.T0051.001` LLM Prompt Injection (Indirect), `AML.T0070`, `AML.T0020`.

### Lifecycle Stages (Optional)

- [ ] planning
- [ ] data-preparation
- [ ] model-training
- [ ] development
- [ ] evaluation
- [ ] deployment
- [ ] runtime
- [ ] maintenance

### Impact Types (Optional)

- [ ] confidentiality
- [ ] integrity
- [ ] availability
- [ ] privacy
- [ ] safety
- [ ] compliance
- [ ] fairness
- [ ] accountability
- [ ] reliability
- [ ] transparency

### Actor Access Levels (Optional)

- [ ] external
- [ ] api
- [ ] user
- [ ] privileged
- [ ] agent
- [ ] supply-chain
- [ ] infrastructure-provider
- [ ] service-provider
- [ ] physical

### Submission Checklist

- [x] I have reviewed existing risks to ensure this is not a duplicate
- [x] I have read the Submission Readiness Guide and believe my proposal meets the quality bar

**Comment by vgudur-dev** [comment]
+1 to adding this as its own risk — the read-after-write-across-turns framing is exactly the right boundary, and it is the gap most practitioner checklists miss.

Three additions that may help it through the content-reviewer, written against the [Submission Readiness Guide](https://github.com/cosai-oasis/secure-ai-tooling/blob/main/risk-map/docs/contributing/submission-readiness-guide.md):

**1. Category and classical equivalent.** The store is written by the agent at runtime, so `risksRuntimeDataSecurity` seems a closer fit than `risksSupplyChainAndDevelopment` (the supply-chain framing fits the repository-file variant but not memory in general). The classical equivalent the guide asks for: tampering with a trusted configuration/state store — once a bad value is persisted every later read trusts it. The AI-specific amplifier is that memory is written from conversation (user, tool output, web content, peer agents) and read back as instructions, so one accepted write changes future behaviour, permissions or trust with no code change and no live attacker.

**2. Framework mappings in the pinned canonical form.** The ATLAS IDs in the examples (T0051, T0070, T0020…) are the *vectors*; the technique that names the memory itself is `AML.T0080.000` (AI Agent Context Poisoning: Memory, created 30 Sep 2025, so it is inside the 5.0.1 pin):

```
mitre-atlas: AML.T0080.000@5.0.1
stride: Tampering
owasp-top10-llm: LLM04:2025
owasp-top10-llm: LLM08:2025
```

**3. External references in the schema shape**, so the examples can cite `{{ref:…}}` sentinels instead of inline URLs:

```yaml
- type: paper
  id: chen-2024-agentpoison
  title: "AgentPoison: Red-teaming LLM Agents via Poisoning Memory or Knowledge Bases"
  url: https://arxiv.org/abs/2407.12784
- type: paper
  id: dong-2025-minja
  title: A Practical Memory Injection Attack against LLM Agents
  url: https://arxiv.org/abs/2503.03704
- type: other
  id: rehberger-2024-spaiware
  title: "Spyware Injection Into Your ChatGPT's Long-Term Memory (SpAIware)"
  url: https://embracethered.com/blog/posts/2024/chatgpt-macos-app-persistent-data-exfiltration/
- type: spec
  id: owasp-agentic-top10-2026
  title: OWASP Top 10 for Agentic Applications 2026 (ASI06 Memory and Context Poisoning)
  url: https://genai.owasp.org/resource/owasp-top-10-for-agentic-applications-for-2026/
```

SpAIware is worth adding to the examples: a disclosed production vulnerability (fixed September 2024) where a web page planted a persistent memory that exfiltrated later conversations — the same mechanism as AOC-10, in a shipped product.

Disclosure: I lead OWASP Agent Memory Guard, an open-source Incubator project that implements write-path controls for this risk, and I joined WS4 this week. Happy to draft the `risks.yaml` entry as a PR (iCLA signed) once the proposal is accepted, with the wording above as a starting point.

**Comment by davidlabianca** [comment]
## How #524 landed: `riskAgentMemoryPoisoning`

- **Issue:** #524, Memory Poisoning
- **Author:** @jo5iah
- **Reviewer:** @davidlabianca
- **Date:** 2026-09-23
- **Disposition:** accepted, reworked

@jo5iah, thank you. The gap is real. The corpus had no risk for an agent's own persistent memory being poisoned, and three demonstrated attacks show it. The entry is on a review branch built on #507's head as it stands. It came out of three commits (`df13f15`, `5416646`, `0e9601d`). This risk is in the first one.

I'd like your review before I open a formal PR. I'll fold it in and do a language pass on the whole branch at that point, so none of the wording below is final.

> **A note on case numbers:** the issue cites the *Agents of Chaos* cases as `AOC-NN`. The corpus uses `CS<N>`, the paper's own "Case Study #N" numbering, so AOC-10 is **CS10** here.

### What landed

```yaml
- id: riskAgentMemoryPoisoning
  title: Agent Memory Poisoning
  category: risksRuntimeDataSecurity
  personas: [personaPlatformProvider, personaAgenticProvider, personaApplicationDeveloper, personaEndUser]
  controls:
    - controlMemoryReferentRevalidation   # new, from #527
    - controlAgentPluginPermissions
    - controlResourceAuthorizationEnforcement
    - controlUntrustedContextContainment
  mappings:
    mitre-atlas: [AML.T0080.000@5.0.1]
    stride: [Tampering]
  lifecycleStage: [deployment, runtime]
  impactType: [integrity, confidentiality, safety]
  actorAccess: [external, user, agent]
```

Full entry: https://github.com/davidlabianca/secure-ai-tooling/blob/review/jo5iah-2026-09-23/risk-map/yaml/risks.yaml#L5811-L5967

<details>
<summary>shortDescription and the scope paragraph</summary>

```yaml
shortDescription:
  - >
    An agent's persistent memory — the store it writes to in one turn and reads back,
    trusted, in a later one — is corrupted so that a malicious instruction, fabricated
    fact, or mutable reference outlives the session that planted it and steers the
    agent's future behavior.
longDescription:
  # ...
  - >-
    The defining property is read-after-write-across-turns: a record planted in one session is
    read back, trusted, in a later one, so the implant outlives the interaction that created
    it — a diagnostic for why memory is the interesting locus, not the boundary of the risk,
    which is scoped to {{componentMemory}}, the agent's own persistent record of its
    interactions and state, not every store an agent happens to write and later read.
```

</details>

### What changed from the proposal, and why

| Field | Proposal | Landed | Why |
|---|---|---|---|
| Title | Memory Poisoning | **Agent Memory Poisoning** | IDs derive from titles. "Memory Poisoning" would give `riskMemoryPoisoning`, not the id you used. The new title gives `riskAgentMemoryPoisoning` and also avoids reading as process memory. |
| Scope | any store written in one turn and read back later, whatever the substrate (vector store, scratchpad, DB row, instruction file, repo file) | **`componentMemory`**: the agent's own persistent record of its interactions and state | Scoped by substrate, the entry would absorb `riskRetrievalVectorStorePoisoning`, `riskPromptResponseCachePoisoning` and parts of the supply-chain risks. Read-after-write-across-turns stays as the property that makes memory dangerous, not as the scope boundary. |
| Category | `risksSupplyChainAndDevelopment` | **`risksRuntimeDataSecurity`** | The agent writes the store at runtime, which matches its sibling `riskPromptResponseCachePoisoning`. The supply-chain framing fits only the repository-file variant. |
| Distinctness | "no entry names `componentMemory` as the poisoned asset" | a paragraph that separates it from five neighbours on substance | Five controls already list `componentMemory`, so the component alone could not carry the argument. The entry now distinguishes itself from `riskRetrievalVectorStorePoisoning` (curated grounding corpora vs the agent's own record: a difference of **role**, not substrate), `riskPromptResponseCachePoisoning`, `riskPromptInjection`, `riskLongLivedSessionStateWeakness` and `riskToolSourceProvenance`. |
| CS10 (AOC-10) | a third incident alongside MINJA and AgentPoison | **the memory-as-pointer variant** | In CS10 the payload was never in memory. The memory file held a *link* to an externally editable document, and a non-owner edited the document afterwards. That makes it a time-of-check-to-time-of-use failure: the write-time check validated the reference, not what it later resolved to. The entry now describes **two** variants: direct content (MINJA, AgentPoison, SpAIware) and memory-as-pointer (CS10). #527's control covers the second one. |
| Examples | MINJA and AgentPoison via the WS2 incident-response doc, each tagged with ATLAS vector IDs | cited to the primary papers through `externalReferences`, plus a SpAIware production example | Examples cite primary sources through `{{ref:…}}`. The per-example ATLAS vector IDs (T0051, T0070, T0020…) are entry vectors, not the memory technique itself, so the entry maps the technique once in `mappings` instead. |
| Evidence register | "three catalogued incidents" | "demonstrated, not merely hypothesized" | MINJA and AgentPoison are research constructions and CS10 is a controlled study, so the wording now says that. SpAIware is the one disclosed production case. |
| Controls | a new `controlAgentMemoryIntegrity` | the narrowed #527 control plus three existing controls | Each control is here because it mitigates the risk. Listing `componentMemory` was not enough to earn a place. Two controls that list the component (`controlIsolatedConfidentialComputing`, `controlAgentCredentialIsolation`) were deliberately left off because neither does anything about a legitimate-channel write. |
| Personas | `personaAgenticProvider`, `personaEndUser` | adds `personaPlatformProvider` and `personaApplicationDeveloper` | These are the personas affected wherever memory is hosted or wired into an application. |

### @vgudur-dev's proposals (2026-09-14)

Thank you. Several of these went straight in. Your framing of the classical equivalent (tampering with a trusted state store, where every later read trusts a value accepted once) is close to how the entry's first paragraph now opens.

| # | Proposal | Disposition |
|---|---|---|
| 1 | `mitre-atlas: AML.T0080.000@5.0.1` | **Accepted.** Checked against the pinned 5.0.1 release: "Memory", a sub-technique of AI Agent Context Poisoning. It is the entry's only ATLAS value. |
| 1 | `stride: Tampering` | **Accepted.** |
| 1 | `owasp-top10-llm: LLM04:2025` | **Not taken.** LLM04's own text is scoped to pre-training, fine-tuning and embedding data. It says nothing about runtime conversational memory, and the entry argues it is *not* training-data poisoning. In the corpus only `riskDataPoisoning`, the entry LLM04 actually describes, carries it. |
| 1 | `owasp-top10-llm: LLM08:2025` | **Not taken here.** Vector and embedding weaknesses would reopen the boundary with retrieval poisoning that the scoping closed. The same review did add LLM08 to `riskRetrievalVectorStorePoisoning`, where it fits: https://github.com/davidlabianca/secure-ai-tooling/blob/review/jo5iah-2026-09-23/risk-map/yaml/risks.yaml#L2468-L2551 |
| 2 | `risksRuntimeDataSecurity` | **Accepted**, as in the table above. |
| 3 | `chen-2024-agentpoison`, `dong-2025-minja` | **Accepted.** MINJA's title is updated to its current arXiv metadata ("Memory Injection Attacks on LLM Agents via Query-Only Interaction"). |
| 3 | `rehberger-2024-spaiware`, plus SpAIware as an example | **Accepted** as `type: advisory`, and it is now the entry's production example. |
| 3 | "the same mechanism as AOC-10" | **Not taken.** SpAIware wrote the payload into memory, which is the direct-content variant. CS10 stored a pointer to content that changed later, which is the pointer variant. They are two mechanisms, and keeping them apart is what justifies #527's control. |
| 3 | `owasp-agentic-top10-2026` (ASI06) | **Deferred.** We have not been able to retrieve and review the normative text, and citing the list's label alone would cite a summary. We will re-evaluate once the text is published. |
| 4 | Offer to draft the `risks.yaml` entry as a PR | Thank you. The entry is already drafted on the branch above, and it uses your input. The most useful help now is review of that text, here or on the formal PR. |

### What I'd like from you

1. Does the `componentMemory` scoping cover the cases the Telemetry SIG had in mind? In particular, does it cover the project-instruction-file and repository-file cases, or do those need a home elsewhere?
2. Is the direct-content / memory-as-pointer split right as written?

The three commits: https://github.com/davidlabianca/secure-ai-tooling/compare/1a0b5d4...review/jo5iah-2026-09-23
